bfvmm/src/hve/arch/intel_x64/ept/memory_map.cpp

```diff
@@ -0,0 +1,340 @@
+//
+// Bareflank Extended APIs
+//
+// Copyright (C) 2015 Assured Information Security, Inc.
+//
+// This library is free software; you can redistribute it and/or
+// modify it under the terms of the GNU Lesser General Public
+// License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+//
+// This library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+// Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public
+// License along with this library; if not, write to the Free Software
+// Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
+
+#include <bfvmm/memory_manager/memory_manager.h>
+#include "hve/arch/intel_x64/ept/memory_map.h"
+#include "hve/arch/intel_x64/ept/intrinsics.h"
+
+namespace eapis
+{
+namespace intel_x64
+{
+namespace ept
+{
+
+memory_map::memory_map()
+{
+    epte_t *pml4 = new epte_t[page_table::num_entries] {0};
+    m_pml4_hva = reinterpret_cast<hva_t>(pml4);
+    m_pml4_hpa = g_mm->virtint_to_physint(m_pml4_hva);
+}
+
+memory_map::~memory_map()
+{
+    epte_t *pml4 = reinterpret_cast<epte_t *>(m_pml4_hva);
+    auto &&pml4_view = gsl::make_span(pml4, page_table::num_entries);
+    for (auto pml4e : pml4_view) {
+        if (epte::is_present(pml4e) && !epte::is_leaf_entry(pml4e)) {
+            free_page_table(pml4e);
+        }
+        epte::clear(pml4e);
+    }
+    delete pml4;
+}
+
+epte_t &
+memory_map::map(gpa_t gpa, hpa_t hpa, uint64_t size) const
+{
+    std::lock_guard<std::mutex> guard(m_mutex);
+
+    switch (size) {
+        case pdpte::page_size_bytes:
+            expects(pdpte::page_address::is_aligned(hpa));
+            return map_pdpte_to_page(gpa, hpa);
+
+        case pde::page_size_bytes:
+            expects(pde::page_address::is_aligned(hpa));
+            return map_pde_to_page(gpa, hpa);
+
+        case pte::page_size_bytes:
+            expects(pte::page_address::is_aligned(hpa));
+            return map_pte_to_page(gpa, hpa);
+
+        default:
+            throw std::logic_error("map: invalid ept page size specified");
+    }
+}
+
+void
+memory_map::unmap(gpa_t gpa) const
+{
+    std::lock_guard<std::mutex> guard(m_mutex);
+
+    epte_t &leaf = gpa_to_epte(gpa);
+    epte::clear(leaf);
+}
+
+epte_t &
+memory_map::gpa_to_epte(gpa_t gpa) const
+{
+    auto err = "gpa_to_epte: failed to resolve gpa->epte, gpa is not mapped";
```

Comment:
Get rid of this, and make per error messages. A general error for the entire function is not descriptive.

Reply:
This error is the same in every place that it is thrown. I don't really know how to be more specific with this message for each case, the entry is either in the map or it isn't. I guess I could add the given gpa to the error message?

Reply:
Its not though as you are erroring out at different levels in the tree.